cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java

```diff
@@ -1135,8 +1175,43 @@ private String getHiveDatabase(@Nullable String namespace) {
     if (namespace == null) {
       return null;
     }
-    String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
-    return namespace.equals(Id.Namespace.DEFAULT.getId()) ? namespace : String.format("%s_%s", tablePrefix, namespace);
+    try {
+      if (namespace.equals(NamespaceId.DEFAULT.getNamespace())) {
+        return namespace;
+      }
+      String hiveDatabase = getCustomHiveDatabaseIfExists(namespace);
+      if (!Strings.isNullOrEmpty(hiveDatabase)) {
+        return hiveDatabase;
+      }
+      String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
+      return String.format("%s_%s", tablePrefix, namespace);
+
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Return the custom database for a given namespace if one exists else return null
+   *
+   * @param namespace the namespace for which we need to look up custom mapping
+   * @return the custom hive database name if one exists else null
+   * @throws ExploreException if failed to get NamespaceMeta for the given namespace
+   */
+  @Nullable
+  private String getCustomHiveDatabaseIfExists(@Nullable String namespace) throws ExploreException {
+    // null namespace implies that the operation happens across all databases
+    // also currently we don't support custom mapping for default namespace
+    if (namespace == null || namespace.equals(NamespaceId.DEFAULT.getNamespace())) {
+      return null;
+    }
+    try {
+      String hiveDatabase = namespaceQueryAdmin.get(Id.Namespace.from(namespace)).getConfig().getHiveDatabase();
+      return Strings.isNullOrEmpty(hiveDatabase) ? null : hiveDatabase;
```

Comment:
so this can never return empty string right? Why all the empty string checks above then?